### PR TITLE
Refactor; Add features to OrienteeringRouter

### PR DESCRIPTION
**Pull request by squarespirit**

### CODE REVIEW

File: planner/routers/point2point_router.py [hunk 1]
```diff
@@ -1,13 +1,48 @@
-from routers.base_router import BaseRouter
+import itertools
+from pprint import pprint
+from typing import *
 
+from routers.base_router import RouteResult
 
-class Point2PointRouter(BaseRouter):
+
+class Point2PointRouter:
```
Comment (veniversum) [inline planner/routers/point2point_router.py]: Use new style classes

File: planner/start_server.py [hunk 2]
```diff
@@ -15,44 +17,64 @@
...
-        origin = constraints.get('origin')
-        destination = constraints.get('destination')
-        desired_length = constraints.get('desiredLength')
-        if origin is not None and destination is not None:
-            orig_lat = float(origin.get('latitude'))
-            orig_lng = float(origin.get('longitude'))
-            dest_lat = float(destination.get('latitude'))
-            dest_lng = float(destination.get('longitude'))
+        req = json.loads(request.jsonData)
+        req = req['jsonData']
```
Comment (veniversum) [inline planner/start_server.py]: Invalid key?

File: planner/start_server.py [hunk 3]
```diff
@@ -15,44 +17,64 @@
...
+
+        try:
+            # Convert origins and dests to lists of tuples
+            origins = [(o['latitude'], o['longitude'])
+                       for o in req['origins']]
+            dests = [(d['latitude'], d['longitude'])
+                     for d in req['dests']]
+
+            # Neither can be empty
+            if origins == [] or dests == []:
```
Comment (veniversum) [inline planner/start_server.py]: Can be simplified to `if not (origins and dests)`

File: planner/start_server.py [hunk 4]
```diff
@@ -15,44 +17,64 @@
...
 
-                linestring, length = router.make_route((orig_lat, orig_lng), (dest_lat, dest_lng), desired_length)
-                route_geometry = json.loads(linestring)
+                else:
+                    length = req['desired_dist']
+                    poi_prefs = req['poi_prefs']
+                    edge_prefs = req['edge_prefs']
+                    noptions = req['noptions']
+                    router = OrienteeringRouter(conn)
+                    routes = router.make_route(
```
Comment (veniversum) [inline planner/start_server.py]: We should have a base class that takes origin, destination and kwargs of options.

File: planner/start_server.py [hunk 5]
```diff
@@ -15,44 +17,64 @@
 _ONE_DAY_IN_SECONDS = 60 * 60 * 24
 
 
+def routeresults_to_json(list: List[RouteResult]) -> str:
```
Comment (veniversum) [inline planner/start_server.py]: Non canonical style, subclass `JSONEncoder` and pass it into `json.dumps` instead.

File: planner/routers/base_router.py [hunk 6]
```diff
@@ -1,8 +1,7 @@
-import abc
+from typing import NamedTuple
 
 
-class BaseRouter(abc.ABC):
-
-    @abc.abstractmethod
-    def make_route(self, origin, destination, *args, **kwargs):
-        raise NotImplementedError
+class RouteResult(NamedTuple):
```
Comment (veniversum) [inline planner/routers/base_router.py]: Why subclass NamedTuple? Not cannonical, see https://docs.python.org/3/library/collections.html#collections.namedtuple.

Also just an object is okay

File: planner/routers/point2point_router.py [hunk 7]
```diff
@@ -1,13 +1,48 @@
...
-                        (*origin[::-1], *destination[::-1]))
-            linestring, length = cur.fetchone()
-            return linestring, length
+            for origin, dest in itertools.product(origins, dests):
+                cur.execute(
+                    'SELECT * FROM pathFromNearestKnownPoints(%s,%s,%s,%s)',
+                    (*origin[::-1], *dest[::-1]))
+                linestring, length = cur.fetchone()
+                output.append(RouteResult(
+                    route=linestring,
```
Comment (veniversum) [inline planner/routers/point2point_router.py]: This doesn't tell you which origins/destinations are used in the routing?

**Review by veniversum — CHANGES_REQUESTED** [review]

**Comment by veniversum** [comment]
Also, use the `logging` module and not print statements.

File: planner/routers/point2point_router.py (hunk 1 above)
Comment (squarespirit) [reply planner/routers/point2point_router.py]: I thought all python 3 classes were new-style? http://portingguide.readthedocs.io/en/latest/classes.html#new-style-classes

File: planner/start_server.py (hunk 4 above)
Comment (squarespirit) [reply planner/start_server.py]: Ok, but if Point2Pointrouter and OrienteeringRouter inherit from it, Pycharm still tells me that the function signatures don't match.

File: planner/routers/base_router.py (hunk 6 above)
Comment (squarespirit) [reply planner/routers/base_router.py]: This is pretty tricky...https://stackoverflow.com/questions/5906831/serializing-a-python-namedtuple-to-json

File: planner/start_server.py (hunk 5 above)
Comment (squarespirit) [reply planner/start_server.py]: This is pretty tricky...https://stackoverflow.com/questions/5906831/serializing-a-python-namedtuple-to-json

File: planner/routers/base_router.py (hunk 6 above)
Comment (squarespirit) [reply planner/routers/base_router.py]: https://docs.python.org/3/library/typing.html#typing.NamedTuple

File: planner/routers/point2point_router.py (hunk 7 above)
Comment (squarespirit) [reply planner/routers/point2point_router.py]: Yeah it doesn't. Though if they were to be specified how would it be done?

File: planner/start_server.py (hunk 4 above)
Comment (veniversum) [reply planner/start_server.py]: Pycharm acts wonky sometimes. It's okay to ignore that inspection warning.

Comment (squarespirit) [reply planner/start_server.py]: I still think it might not make sense. The make_route call is not polymorphic.

Comment (veniversum) [reply planner/start_server.py]: I feel that the logic of parsing options should be in the route planners, that way it's easier to have other routers.

File: planner/routers/point2point_router.py (hunk 7 above)
Comment (veniversum) [reply planner/routers/point2point_router.py]: Maybe start vertex and end vertex, indexed by the origin/destination lists?

For example one-to-one routing will have both start and end vertex id as 0.

File: planner/start_server.py (hunk 4 above)
Comment (squarespirit) [reply planner/start_server.py]: But you need to look at the options to decide which route planner to use.

File: planner/routers/point2point_router.py (hunk 7 above)
Comment (squarespirit) [reply planner/routers/point2point_router.py]: Technically you can get that by looking at the start and end of the linestring...I kind of want to wait until frontend/web API tells us exactly what they want before implementing it?

Comment (squarespirit) [reply planner/routers/point2point_router.py]: We also don't return a list of the POIs visited on each route, nor any information about the POIs.

Comment (veniversum) [reply planner/routers/point2point_router.py]: Yeah but we have a 'snap to grid' feature in the route planner, so the actual start and end may not be the same points as the origins/destinations provided.

**Pull request merged**